### PR TITLE
feat(collection): add findSingle

### PR DESCRIPTION
**Pull request by NotFounds**
Implement `single` for collection module.
(ref: https://github.com/denoland/deno_std/issues/1065)

The type definition is based on [this](https://github.com/denoland/deno_std/pull/978/files#diff-d4cdef3eccda16457c371d681c99b711821c5ad7f436a31e852f44a7142bcf72R242-R258).

ref. 
https://github.com/denoland/deno_std/issues/1065#issuecomment-906123863
https://github.com/denoland/deno_std/pull/978

### CODE REVIEW

File: collections/README.md [hunk 1]
```diff
@@ -395,6 +395,27 @@ console.assert(
 );
 ```
 
+### single
+
+Returns the only element in the given collection matching the given predicate.
+Returns undefined if there is none or multiple matches for the predicate.
```
[the remark this answers is not in the document]
Comment (NotFounds) [reply collections/README.md]: Thank you for comment! Fixed it! 1fa2dbb

**Comment by LionC** [comment]
@kt3k I think this is approaching mergeability - what should we do about the name?

**Comment by kt3k** [comment]
@LionC 
> what should we do about the name?

I started feeling the name is ok as is because `single` seems having noun usage.

Does anyone have any opinion about this? cc @timreichen

**Comment by timreichen** [comment]
@kt3k
I think `single` is not a good name, firstly because it is not a verb and secondly because it is not descriptive at all on what it does.
Since we have a `distinct` and `distinctBy`, how about `findDistinct`?

**Comment by LionC** [comment]
> @kt3k
> I think `single` is not a good name, firstly because it is not a verb and secondly because it is not descriptive at all on what it does.
> Since we have a `distinct` and `distinctBy`, how about `findDistinct`?

If you read

```ts
const file = findDistinct(Deno.args)
```

do you feel like you would know what's going on here?

In my (non-native speaker) mind the word distinct needs other values to make sense and does not really make sense for a one-element array.

**Comment by kt3k** [comment]
@timreichen Thank you for the input, but findDistinct sounds a little strange to me..

How about `findSingle`?

**Comment by timreichen** [comment]
@kt3k This will make deno std attractive for people that do partnership search:) I don't know, but better than just `single` still not really expressive.
@LionC
```ts
const file = findDistinct(Deno.args)
```
tbh not so much since the default predicate is set to `(_) => true` (which I don't know is a good idea *if* we use `find` as the verb as  in js `find` and `findIndex` have a mandatory predicates?).

In that sense I think having this default predicate changes the meaning of the function to `Deno.args.length === 1 ? Deno.args[0] : undefined` where with a custom predicate it is about finding a distinct value in an array.

```ts
const value = findDistinct(["f", "o", "o"], (char) => char === "o") // undefined
```
```ts
const value = findDistinct(["f", "o", "o"], (char) => char === "f") // "f"
```

btw:
not sure about this check:
```ts
if (match !== undefined) return undefined;`
```
Hypothetically this will not short circuit for undefined values:
```ts
const array = new Array(1000).fill(undefined)
const value = findDistinct(array, (i) => i === undefined);
```

Maybe a flag should be used instead?
```ts
let match: T | undefined = undefined;
let found = false
for (const element of array) {
  if (predicate(element)) {
    if (found) return undefined;
    found = true
    match = element;
  }
}
return match;
```

**Comment by LionC** [comment]
@timreichen 

> @kt3k This will make deno std attractive for people that do partnership search:)

Adoption boost spotted^^

> tbh not so much since the default predicate is set to `(_) => true` (which I don't know is a good idea *if* we use `find` as the verb as  in js `find` and `findIndex` have a mandatory predicates?).

Yeah that is what I meant - find does not fit that well but it is an essential part of the usage/signature (stolen from kotlin again).

> Hypothetically this will not short circuit for undefined values:
> 
> ```ts
> 
> const array = new Array(1000).fill(undefined)
> 
> const value = findDistinct(array, (i) => i === undefined); // predicate called 1000x
> 
> ```

Oh yes, we should change that @NotFounds . Good catch!

**Comment by NotFounds** [comment]
@timreichen 
Thank you for your review and suggestion.
I didn't take that case into consideration...

Fixed it! 8864057

**Comment by LionC** [comment]
Should be mergeable now @kt3k

**Comment by kt3k** [comment]
Can we settle on the name 'single'?

@NotFounds Do you have any preference on naming?

**Comment by NotFounds** [comment]
(I'm not a native speaker)
I'm comfortable with the use of "single".
And especially when the behavior is the same as the method used in other languages.

But... if a verb is preferred, the suggested "findSingle" is the best fit for me.

**Comment by lucacasonato** [comment]
`takeSingle`?

**Comment by LionC** [comment]
Sounds like `single` and `findSingle` are the options. I would vote for `single` because it is short :-)

@kt3k I think we need your/some core call here.

**Comment by kt3k** [comment]
Note

NotFounds: single or findSingle
Leon: single
Luca: findSingle
Tim: findDistinct
Grian: single or findSingle
Yoshi: single or findSingle

**Comment by timreichen** [comment]
@kt3k
I would join the findSingle crew if we removed the default selector (reasons why described above)

**Comment by LionC** [comment]
I dont think we should remove functionality for a name, but find a name for functionality

**Comment by timreichen** [comment]
I think `findSingle` describes the algorithm of the function pretty well. So I argue that it is not removing functionality but removing a bad default that changes the intent of the function. Consider `findSingle([undefined, undefined])` and `findSingle([])`: The return values will be `undefined` for both with the current default predicate, making the intended usage of this function questionable. That default predicate might make sense in kotlin because it throws an exception if it is not successful. Because this implementation returns `undefined` on failure that predicate is not really useful imo.

To make sure a single value exists in an array  `array.length === 1` should be used instead of that function anyhow, so why not generalize it by removing the predicate therefore solving the naming problem too?

**Comment by LionC** [comment]
> I think `findSingle` or `findDistinct` describes the algorithm of the function pretty well. So I argue that it is not removing functionality but removing a bad default that changes the intent of the function. Consider `findSingle([])`, `findSingle([undefined])` and `findSingle([undefined, undefined])`: The return values will be `undefined` for all of them with the current default predicate, making the intended usage of this function questionable. That default predicate might make sense in kotlin because it throws an exception if it is not successful. Because this implementation returns `undefined` on failure that predicate is not really useful imo.

Typings make that `undefined` case pretty clear I think - I can not come up with any bugs that could cause, but maybe I am missing them. It is really similar to an index access returning `undefined` vs that index's value actually being `undefined`.
 
> To make sure a single value exists in an array `array.length === 1` should be used instead of that function anyhow, so why not generalize it by removing the predicate therefore solving the naming problem too?

I mean you can always write more code instead of using something that encapsulates it, sure. Saying "should" here is a stretch to me though - why "should" someone not use that function if they want to? It might not be your style, but it might be theirs and it is hard to say someone objectively "should" do not do it.

Either way, I think I would be fine with `findSingle` the way it is right now if that is what it takes to get this merged^^

**Comment by kt3k** [comment]
Let's settle on the name `findSingle` because everybody is at least OK with it.

Let's continue the discussion of the default selector in another issue.

I'm merging this for now as this one is enough discussed and looks having enough quality now. Thank you for your efforts!

**Review by kt3k — APPROVED** [review]
@NotFounds LGTM. Thank you for your contribution!

**Pull request merged**